File: airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/orchestrator/ReplicationJobOrchestrator.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.container_orchestrator.orchestrator;

import static io.airbyte.metrics.lib.ApmTraceConstants.JOB_ORCHESTRATOR_OPERATION_NAME;
import static io.airbyte.metrics.lib.ApmTraceConstants.Tags.DESTINATION_DOCKER_IMAGE_KEY;
import static io.airbyte.metrics.lib.ApmTraceConstants.Tags.JOB_ID_KEY;
import static io.airbyte.metrics.lib.ApmTraceConstants.Tags.SOURCE_DOCKER_IMAGE_KEY;

import com.google.common.annotations.VisibleForTesting;
import datadog.trace.api.Trace;
import io.airbyte.commons.json.Jsons;
import io.airbyte.commons.temporal.TemporalUtils;
import io.airbyte.config.FailureReason;
import io.airbyte.config.ReplicationAttemptSummary;
import io.airbyte.config.ReplicationOutput;
import io.airbyte.featureflag.CoRoutineBufferedReplicationWorker;
import io.airbyte.featureflag.Connection;
import io.airbyte.featureflag.FeatureFlagClient;
import io.airbyte.metrics.MetricAttribute;
import io.airbyte.metrics.MetricClient;
import io.airbyte.metrics.OssMetricsRegistry;
import io.airbyte.metrics.lib.ApmTraceUtils;
import io.airbyte.persistence.job.models.JobRunConfig;
import io.airbyte.persistence.job.models.ReplicationInput;
import io.airbyte.workers.exception.WorkerException;
import io.airbyte.workers.general.BufferedReplicationWorker;
import io.airbyte.workers.general.ReplicationWorkerFactory;
import io.airbyte.workers.general.buffered.worker.ReplicationWorkerK;
import io.airbyte.workers.helper.FailureHelper;
import io.airbyte.workers.internal.exception.DestinationException;
import io.airbyte.workers.internal.exception.SourceException;
import io.airbyte.workers.workload.JobOutputDocStore;
import io.airbyte.workload.api.client.WorkloadApiClient;
import io.airbyte.workload.api.client.model.generated.WorkloadCancelRequest;
import io.airbyte.workload.api.client.model.generated.WorkloadFailureRequest;
import io.airbyte.workload.api.client.model.generated.WorkloadSuccessRequest;
import jakarta.inject.Named;
import jakarta.inject.Singleton;
import java.io.IOException;
import java.lang.invoke.MethodHandles;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.ToDoubleFunction;
import kotlin.coroutines.EmptyCoroutineContext;
import kotlinx.coroutines.BuildersKt;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Runs replication worker.
 */
@Singleton
public class ReplicationJobOrchestrator {

  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

  @VisibleForTesting
  public static final double BYTES_TO_GB = 1024 * 1024 * 1024;
  @VisibleForTesting
  public static final String STATUS_ATTRIBUTE = "status";
  @VisibleForTesting
  public static final String FAILED_STATUS = "failed";
  @VisibleForTesting
  public static final String SUCCESS_STATUS = "success";

  private static final ToDoubleFunction<ReplicationAttemptSummary> BYTES_GAUGE_FUNCTION =
      (replicationAttemptSummary) -> replicationAttemptSummary.getBytesSynced() / BYTES_TO_GB;
  private static final ToDoubleFunction<ReplicationAttemptSummary> DURATION_GAUGE_FUNCTION =
      (replicationAttemptSummary) -> Duration.ofMillis(replicationAttemptSummary.getEndTime() - replicationAttemptSummary.getStartTime()).toSeconds();

  private final ReplicationInput replicationInput;
  private final Path workspaceRoot;
  private final JobRunConfig jobRunConfig;
  private final ReplicationWorkerFactory replicationWorkerFactory;
  private final WorkloadApiClient workloadApiClient;
  private final JobOutputDocStore jobOutputDocStore;
  private final String workloadId;
  private final MetricClient metricClient;
  private final FeatureFlagClient featureFlagClient;

  public ReplicationJobOrchestrator(final ReplicationInput replicationInput,
                                    @Named("workloadId") final String workloadId,
                                    @Named("workspaceRoot") final Path workspaceRoot,
                                    final JobRunConfig jobRunConfig,
                                    final ReplicationWorkerFactory replicationWorkerFactory,
                                    final WorkloadApiClient workloadApiClient,
                                    final JobOutputDocStore jobOutputDocStore,
                                    final MetricClient metricClient,
                                    final FeatureFlagClient featureFlagClient) {
    this.replicationInput = replicationInput;
    this.workloadId = workloadId;
    this.workspaceRoot = workspaceRoot;
    this.jobRunConfig = jobRunConfig;
    this.replicationWorkerFactory = replicationWorkerFactory;
    this.workloadApiClient = workloadApiClient;
    this.jobOutputDocStore = jobOutputDocStore;
    this.metricClient = metricClient;
    this.featureFlagClient = featureFlagClient;
  }

  @Trace(operationName = JOB_ORCHESTRATOR_OPERATION_NAME)
  public Optional<String> runJob() throws Exception {

    final var sourceLauncherConfig = replicationInput.getSourceLauncherConfig();

    final var destinationLauncherConfig = replicationInput.getDestinationLauncherConfig();

    ApmTraceUtils.addTagsToTrace(
        Map.of(JOB_ID_KEY, jobRunConfig.getJobId(),
            DESTINATION_DOCKER_IMAGE_KEY, destinationLauncherConfig.getDockerImage(),
            SOURCE_DOCKER_IMAGE_KEY, sourceLauncherConfig.getDockerImage()));
    boolean shouldUseNewReplicationWorker =
        featureFlagClient.boolVariation(CoRoutineBufferedReplicationWorker.INSTANCE, new Connection(replicationInput.getConnectionId()));

    log.info("Running replication worker...");
    final var jobRoot = TemporalUtils.getJobRoot(workspaceRoot,
        jobRunConfig.getJobId(), jobRunConfig.getAttemptId());

    final ReplicationOutput replicationOutput = shouldUseNewReplicationWorker
        ? run(replicationWorkerFactory.createK(replicationInput, jobRunConfig, sourceLauncherConfig, destinationLauncherConfig, () -> {}, workloadId),
            replicationInput, jobRoot, workloadId)
        : run(replicationWorkerFactory.create(replicationInput, jobRunConfig, sourceLauncherConfig, destinationLauncherConfig, () -> {}, workloadId),
            replicationInput, jobRoot, workloadId);

    jobOutputDocStore.writeSyncOutput(workloadId, replicationOutput);
    updateStatusInWorkloadApi(replicationOutput, workloadId);

    final List<MetricAttribute> attributes = buildMetricAttributes(replicationInput,
        Long.parseLong(jobRunConfig.getJobId()), jobRunConfig.getAttemptId().intValue());
    metricClient.gauge(OssMetricsRegistry.SYNC_DURATION,
        replicationOutput.getReplicationAttemptSummary(),
        DURATION_GAUGE_FUNCTION,
        attributes.toArray(new MetricAttribute[0]));
    metricClient.gauge(OssMetricsRegistry.SYNC_GB_MOVED,
        replicationOutput.getReplicationAttemptSummary(),
        BYTES_GAUGE_FUNCTION,
        attributes.toArray(new MetricAttribute[0]));

    log.info("Returning output...");
    return Optional.of(Jsons.serialize(replicationOutput));
  }

  private ReplicationOutput run(final ReplicationWorkerK replicationWorker,
                                final ReplicationInput replicationInput,
                                final Path jobRoot,
                                final String workloadId)
      throws InterruptedException {

    return BuildersKt.runBlocking(
        EmptyCoroutineContext.INSTANCE,
        (scope, continuation) -> {
          final Long jobId = Long.parseLong(jobRunConfig.getJobId());
          final Integer attemptNumber = Math.toIntExact(jobRunConfig.getAttemptId());
          final List<MetricAttribute> attributes = buildMetricAttributes(replicationInput, jobId, attemptNumber);
          try {
            return replicationWorker.run(replicationInput, jobRoot, continuation);
          } catch (final DestinationException e) {
            try {
              failWorkload(workloadId, Optional.of(FailureHelper.destinationFailure(e, jobId, attemptNumber)));
            } catch (final IOException ioe) {
              e.addSuppressed(ioe);
            }
            attributes.add(new MetricAttribute(STATUS_ATTRIBUTE, FAILED_STATUS));
            throw e;
          } catch (final SourceException e) {
            try {
              failWorkload(workloadId, Optional.of(FailureHelper.sourceFailure(e, jobId, attemptNumber)));
            } catch (final IOException ioe) {
              e.addSuppressed(ioe);
            }
            attributes.add(new MetricAttribute(STATUS_ATTRIBUTE, FAILED_STATUS));
            throw e;
          } catch (final WorkerException e) {
            try {
              failWorkload(workloadId, Optional.of(FailureHelper.platformFailure(e, jobId, attemptNumber)));
            } catch (final IOException ioe) {
              e.addSuppressed(ioe);
            }
            attributes.add(new MetricAttribute(STATUS_ATTRIBUTE, FAILED_STATUS));
            throw new RuntimeException(e);
          } finally {
            if (attributes.stream().noneMatch(a -> STATUS_ATTRIBUTE.equalsIgnoreCase(a.getKey()))) {
              attributes.add(new MetricAttribute(STATUS_ATTRIBUTE, SUCCESS_STATUS));
            }
            metricClient.count(OssMetricsRegistry.SYNC_STATUS, attributes.toArray(new MetricAttribute[0]));
          }
        });
  }

  @VisibleForTesting
  ReplicationOutput run(final BufferedReplicationWorker replicationWorker,
                        final ReplicationInput replicationInput,
                        final Path jobRoot,
                        final String workloadId)
      throws WorkerException, IOException {

    final Long jobId = Long.parseLong(jobRunConfig.getJobId());
    final Integer attemptNumber = Math.toIntExact(jobRunConfig.getAttemptId());
    final List<MetricAttribute> attributes = buildMetricAttributes(replicationInput, jobId, attemptNumber);

    try {
      return replicationWorker.run(replicationInput, jobRoot);
    } catch (final DestinationException e) {
      failWorkload(workloadId, Optional.of(FailureHelper.destinationFailure(e, jobId, attemptNumber)));
      attributes.add(new MetricAttribute(STATUS_ATTRIBUTE, FAILED_STATUS));
      throw e;
    } catch (final SourceException e) {
      failWorkload(workloadId, Optional.of(FailureHelper.sourceFailure(e, jobId, attemptNumber)));
      attributes.add(new MetricAttribute(STATUS_ATTRIBUTE, FAILED_STATUS));
      throw e;
    } catch (final WorkerException e) {
      failWorkload(workloadId, Optional.of(FailureHelper.platformFailure(e, jobId, attemptNumber)));
      attributes.add(new MetricAttribute(STATUS_ATTRIBUTE, FAILED_STATUS));
      throw e;
    } finally {
      if (attributes.stream().noneMatch(a -> STATUS_ATTRIBUTE.equalsIgnoreCase(a.getKey()))) {
        attributes.add(new MetricAttribute(STATUS_ATTRIBUTE, SUCCESS_STATUS));
      }
      metricClient.count(OssMetricsRegistry.SYNC_STATUS, attributes.toArray(new MetricAttribute[0]));
    }
  }

  @VisibleForTesting
  void updateStatusInWorkloadApi(final ReplicationOutput replicationOutput, final String workloadId) throws IOException {
    if (replicationOutput == null || replicationOutput.getReplicationAttemptSummary() == null) {
      log.warn("The replication output is null, skipping updating the workload status via API");
      return;
    }
    switch (replicationOutput.getReplicationAttemptSummary().getStatus()) {
      case FAILED -> failWorkload(workloadId, replicationOutput.getFailures().stream().findFirst());
      case CANCELLED -> cancelWorkload(workloadId);
      case COMPLETED -> succeedWorkload(workloadId);
      default -> throw new RuntimeException(String.format("Unknown status %s.", replicationOutput.getReplicationAttemptSummary().getStatus()));
    }
  }

  private void cancelWorkload(final String workloadId) throws IOException {
    workloadApiClient.getWorkloadApi().workloadCancel(new WorkloadCancelRequest(workloadId, "Replication job has been cancelled", "orchestrator"));
  }

  private void failWorkload(final String workloadId, final Optional<FailureReason> failureReason) throws IOException {
    if (failureReason.isPresent()) {
      workloadApiClient.getWorkloadApi().workloadFailure(new WorkloadFailureRequest(workloadId,
          failureReason.get().getFailureOrigin().value(),
          failureReason.get().getExternalMessage()));
    } else {
      workloadApiClient.getWorkloadApi().workloadFailure(new WorkloadFailureRequest(workloadId, null, null));
    }
  }

  private void succeedWorkload(final String workloadId) throws IOException {
    workloadApiClient.getWorkloadApi().workloadSuccess(new WorkloadSuccessRequest(workloadId));
  }

  @VisibleForTesting
  public List<MetricAttribute> buildMetricAttributes(final ReplicationInput replicationInput, final Long jobId, final Integer attemptNumber) {
    final List<MetricAttribute> attributes = new ArrayList<>();
    attributes.add(new MetricAttribute("attempt_count", String.valueOf(attemptNumber + 1))); // Normalize to make it human understandable
    attributes.add(new MetricAttribute("connection_id", replicationInput.getConnectionId().toString()));
    attributes.add(new MetricAttribute("workspace_id", replicationInput.getWorkspaceId().toString()));
    attributes.add(new MetricAttribute("job_id", jobId.toString()));
    attributes.add(new MetricAttribute("destination_connector_id", replicationInput.getDestinationId().toString()));
    attributes.add(new MetricAttribute("source_connector_id", replicationInput.getSourceId().toString()));
    return attributes;
  }

}

File: airbyte-container-orchestrator/src/main/kotlin/io/airbyte/container/orchestrator/worker/ReplicationJobOrchestrator.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.container.orchestrator.worker

import com.google.common.annotations.VisibleForTesting
import datadog.trace.api.Trace
import io.airbyte.commons.json.Jsons
import io.airbyte.config.FailureReason
import io.airbyte.config.ReplicationAttemptSummary
import io.airbyte.config.ReplicationOutput
import io.airbyte.config.StandardSyncSummary
import io.airbyte.metrics.MetricAttribute
import io.airbyte.metrics.MetricClient
import io.airbyte.metrics.OssMetricsRegistry
import io.airbyte.metrics.lib.ApmTraceConstants
import io.airbyte.metrics.lib.ApmTraceUtils
import io.airbyte.persistence.job.models.JobRunConfig
import io.airbyte.persistence.job.models.ReplicationInput
import io.airbyte.workers.exception.WorkerException
import io.airbyte.workers.helper.FailureHelper.destinationFailure
import io.airbyte.workers.helper.FailureHelper.platformFailure
import io.airbyte.workers.helper.FailureHelper.sourceFailure
import io.airbyte.workers.internal.exception.DestinationException
import io.airbyte.workers.internal.exception.SourceException
import io.airbyte.workers.workload.WorkloadOutputWriter
import io.airbyte.workload.api.client.WorkloadApiClient
import io.airbyte.workload.api.client.model.generated.WorkloadCancelRequest
import io.airbyte.workload.api.client.model.generated.WorkloadFailureRequest
import io.airbyte.workload.api.client.model.generated.WorkloadSuccessRequest
import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.inject.Named
import jakarta.inject.Singleton
import java.io.IOException
import java.nio.file.Path
import java.time.Duration
import java.util.Optional
import java.util.function.ToDoubleFunction

private val logger = KotlinLogging.logger {}

@VisibleForTesting
internal val BYTES_TO_GB: Double = (1024 * 1024 * 1024).toDouble()

@VisibleForTesting
internal const val STATUS_ATTRIBUTE: String = "status"

@VisibleForTesting
internal const val FAILED_STATUS: String = "failed"

@VisibleForTesting
internal const val SUCCESS_STATUS: String = "success"

private val BYTES_GAUGE_FUNCTION =
  ToDoubleFunction { replicationAttemptSummary: ReplicationAttemptSummary -> (replicationAttemptSummary.bytesSynced ?: 0) / BYTES_TO_GB }

private val DURATION_GAUGE_FUNCTION =
  ToDoubleFunction { replicationAttemptSummary: ReplicationAttemptSummary ->
    Duration.ofMillis(replicationAttemptSummary.endTime - replicationAttemptSummary.startTime).toSeconds().toDouble()
  }

/**
 * Runs replication worker.
 */
@Singleton
class ReplicationJobOrchestrator(
  private val replicationInput: ReplicationInput,
  @Named("workloadId") private val workloadId: String,
  @Named("jobRoot") private val jobRoot: Path,
  private val jobRunConfig: JobRunConfig,
  private val replicationWorker: ReplicationWorker,
  private val workloadApiClient: WorkloadApiClient,
  private val outputWriter: WorkloadOutputWriter,
  private val metricClient: MetricClient,
) {
  @Trace(operationName = ApmTraceConstants.JOB_ORCHESTRATOR_OPERATION_NAME)
  @Throws(Exception::class)
  fun runJob(): Optional<String> {
    val sourceLauncherConfig = replicationInput.sourceLauncherConfig
    val destinationLauncherConfig = replicationInput.destinationLauncherConfig

    ApmTraceUtils.addTagsToTrace(
      mutableMapOf<String, Any>(
        ApmTraceConstants.Tags.IS_RESET_KEY to replicationInput.isReset,
        ApmTraceConstants.Tags.JOB_ID_KEY to jobRunConfig.jobId,
        ApmTraceConstants.Tags.DESTINATION_DOCKER_IMAGE_KEY to destinationLauncherConfig.dockerImage,
        ApmTraceConstants.Tags.SOURCE_DOCKER_IMAGE_KEY to sourceLauncherConfig.dockerImage,
      ),
    )

    logger.info { "Running replication worker..." }

    val replicationOutput =
      run(replicationWorker, replicationInput, jobRoot, workloadId)

    outputWriter.writeSyncOutput(workloadId, replicationOutput)
    updateStatusInWorkloadApi(replicationOutput, workloadId)

    val attributes =
      buildMetricAttributes(
        replicationInput,
        jobRunConfig.jobId.toLong(),
        jobRunConfig.attemptId.toInt(),
      )
    metricClient.gauge<ReplicationAttemptSummary>(
      OssMetricsRegistry.SYNC_DURATION,
      replicationOutput.replicationAttemptSummary,
      DURATION_GAUGE_FUNCTION,
      *attributes.toTypedArray<MetricAttribute>(),
    )
    metricClient.gauge<ReplicationAttemptSummary>(
      OssMetricsRegistry.SYNC_GB_MOVED,
      replicationOutput.replicationAttemptSummary,
      BYTES_GAUGE_FUNCTION,
      *attributes.toTypedArray<MetricAttribute>(),
    )

    logger.info { "Returning output..." }
    return Optional.of(Jsons.serialize(replicationOutput))
  }

  private fun run(
    replicationWorker: ReplicationWorker,
    replicationInput: ReplicationInput,
    jobRoot: Path,
    workloadId: String,
  ): ReplicationOutput {
    val jobId = jobRunConfig.jobId.toLong()
    val attemptNumber = Math.toIntExact(jobRunConfig.attemptId)
    val attributes = buildMetricAttributes(replicationInput, jobId, attemptNumber)
    try {
      return replicationWorker.runReplicationBlocking(jobRoot)
    } catch (e: DestinationException) {
      failWorkload(workloadId = workloadId, failureReason = destinationFailure(e, jobId, attemptNumber), originalException = e)
      attributes.add(MetricAttribute(STATUS_ATTRIBUTE, FAILED_STATUS))
      throw e
    } catch (e: SourceException) {
      failWorkload(workloadId = workloadId, failureReason = sourceFailure(e, jobId, attemptNumber), originalException = e)
      attributes.add(MetricAttribute(STATUS_ATTRIBUTE, FAILED_STATUS))
      throw e
    } catch (e: WorkerException) {
      failWorkload(workloadId = workloadId, failureReason = platformFailure(e, jobId, attemptNumber), originalException = e)
      attributes.add(MetricAttribute(STATUS_ATTRIBUTE, FAILED_STATUS))
      throw RuntimeException(e)
    } catch (e: Exception) {
      failWorkload(workloadId = workloadId, failureReason = platformFailure(e, jobId, attemptNumber), originalException = e)
      throw RuntimeException(e)
    } finally {
      if (attributes.stream().noneMatch { a: MetricAttribute? -> STATUS_ATTRIBUTE.equals(a!!.key, ignoreCase = true) }) {
        attributes.add(MetricAttribute(STATUS_ATTRIBUTE, SUCCESS_STATUS))
      }
      metricClient.count(metric = OssMetricsRegistry.SYNC_STATUS, attributes = attributes.toTypedArray<MetricAttribute>())
    }
  }

  @VisibleForTesting
  @Throws(IOException::class)
  fun updateStatusInWorkloadApi(
    replicationOutput: ReplicationOutput?,
    workloadId: String,
  ) {
    if (replicationOutput == null || replicationOutput.replicationAttemptSummary == null) {
      logger.warn { "The replication output is null, skipping updating the workload status via API" }
      return
    }
    when (replicationOutput.replicationAttemptSummary.status) {
      StandardSyncSummary.ReplicationStatus.FAILED -> failWorkload(workloadId, replicationOutput.failures.first())
      StandardSyncSummary.ReplicationStatus.CANCELLED -> cancelWorkload(workloadId)
      StandardSyncSummary.ReplicationStatus.COMPLETED -> succeedWorkload(workloadId)
      else -> throw RuntimeException("Unknown status ${replicationOutput.replicationAttemptSummary.status}.")
    }
  }

  @Throws(IOException::class)
  private fun cancelWorkload(workloadId: String) {
    workloadApiClient.workloadApi.workloadCancel(WorkloadCancelRequest(workloadId, "Replication job has been cancelled", "orchestrator"))
  }

  private fun failWorkload(
    workloadId: String,
    failureReason: FailureReason?,
    originalException: Exception,
  ) {
    try {
      failWorkload(workloadId, failureReason)
    } catch (ioe: IOException) {
      originalException.addSuppressed(ioe)
    }
  }

  @Throws(IOException::class)
  private fun failWorkload(
    workloadId: String,
    failureReason: FailureReason?,
  ) {
    if (failureReason != null) {
      workloadApiClient.workloadApi.workloadFailure(
        WorkloadFailureRequest(
          workloadId,
          failureReason.failureOrigin.value(),
          failureReason.externalMessage,
        ),
      )
    } else {
      workloadApiClient.workloadApi.workloadFailure(WorkloadFailureRequest(workloadId, null, null))
    }
  }

  @Throws(IOException::class)
  private fun succeedWorkload(workloadId: String) {
    workloadApiClient.workloadApi.workloadSuccess(WorkloadSuccessRequest(workloadId))
  }

  @VisibleForTesting
  fun buildMetricAttributes(
    replicationInput: ReplicationInput,
    jobId: Long,
    attemptNumber: Int,
  ) = mutableListOf(
    MetricAttribute("attempt_count", (attemptNumber + 1).toString()), // Normalize to make it human understandable
    MetricAttribute("connection_id", replicationInput.connectionId.toString()),
    MetricAttribute("workspace_id", replicationInput.workspaceId.toString()),
    MetricAttribute("job_id", jobId.toString()),
    MetricAttribute("destination_connector_id", replicationInput.destinationId.toString()),
    MetricAttribute("source_connector_id", replicationInput.sourceId.toString()),
  )
}
